### PR TITLE
Gate reading of optional string array for bwc

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Missing a check on the transport version results in unreadable cluster state
if it includes a serialized instance of `DownsampleShardTaskParams`.
#98023 introduced an optional string array including dimensions used by time
serie indices.
Reading an optional array requires reading a boolean first which is required to
know if an array of values exists in serialized form. From 8.13 on we try to
read such a boolean which is not there because older versions don't write any
boolean nor any string array. Here we include the check on versions for backward
compatibility skipping reading any boolean or array whatsoever whenever not possible.

Customers using downsampling might have cluster states including such serielized
objects and would be unable to upgrade to version 8.13. They will be able to
upgrade to any version including this fix.

This fix has a side effect https://github.com/elastic/elasticsearch/issues/106880

### CODE REVIEW

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java [hunk 1]
```diff
@@ -108,7 +110,10 @@ public void writeTo(StreamOutput out) throws IOException {
         shardId.writeTo(out);
         out.writeStringArray(metrics);
         out.writeStringArray(labels);
-        out.writeOptionalStringArray(dimensions);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.DIMENSIONS_SERIALIZATION_PERSISTENT_TASK)) {
+            out.writeOptionalStringArray(dimensions);
+        }
+        out.writeBoolean(false);
```
Comment (martijnvg) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java]: I think `out.writeBoolean(false);` should be removed?

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java [hunk 2]
```diff
@@ -71,7 +71,9 @@ public record DownsampleShardTaskParams(
             new ShardId(in),
             in.readStringArray(),
             in.readStringArray(),
-            in.readOptionalStringArray()
+            in.getTransportVersion().onOrAfter(TransportVersions.DIMENSIONS_SERIALIZATION_PERSISTENT_TASK)
+                ? in.readOptionalStringArray()
+                : new String[] {}
```
Comment (kkrik-es) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java:76]: What happens when a task with no dimensions gets executed? Shall we update the executor to abort?

Comment (salvatore-campagna) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java:76]: The task is expected to fail with an unretryable error because of no dimensions.

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java [hunk 3]
```diff
@@ -148,6 +148,17 @@ class DownsampleShardIndexer {
 
     public DownsampleIndexerAction.ShardDownsampleResponse execute() throws IOException {
         final Query initialStateQuery = createQuery();
+        if (this.dimensions == null || this.dimensions.length <= 0) {
```
Comment (salvatore-campagna) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java:151]: I expect this to not be really necessary because creating a tsdb index without dimensions is not possible. and that would be the only way I see to have empty dimensions.
Anyway I think adding this check is ok udner the assumption that eventually failing those tasks is not an issue and does not create other issues.

**Comment by salvatore-campagna** [comment]
Note that this issue as a side effect described here https://github.com/elastic/elasticsearch/issues/106880

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java:76]: I think we should not fail, but use the old way of downsampling for tasks without dimensions, but I might be missing something obvious here.

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java (hunk 3 above)
Comment (martijnvg) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java:151]: Tsdb indices created before 8.13.0 can get the dimensions for the the tsid? Since on earlier version the tsid isn't hashed and it contains the dimension field values pairs.

Comment (dnhatn) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java:151]: Yes, I brought back this old logic in https://github.com/elastic/elasticsearch/pull/106878/commits/132e5f12e0a926d706e89230b8485a6e08236234

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import org.elasticsearch.gradle.Version
```
Comment (martijnvg) [inline x-pack/plugin/downsample/qa/mixed-cluster/build.gradle:8]: I think the tests from `./x-pack/plugin/downsample/qa/rest` can be moved to `./x-pack/plugin/srx/yamlRestTest` and that way we would picky bag on the existing mixed cluster qa module? Instead of copying these tests into a new qa module?

I don't think we need to address this here, the important thing is to get this fix. Maybe in a followup?

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java [hunk 5]
```diff
@@ -564,6 +565,18 @@ public XContentBuilder buildDownsampleDocument() throws IOException {
                 fieldProducer.write(builder);
             }
 
+            if (dimensions.length == 0) {
+                if (indexShard.indexSettings().getIndexVersionCreated().onOrAfter(IndexVersions.TIME_SERIES_ID_HASHING)) {
+                    throw new IllegalStateException("downsampling task of indices create in 8.13 or later should specify dimensions");
+                }
+                // load dimensions from the legacy tsid
+                Map<?, ?> dimensions = (Map<?, ?>) DocValueFormat.TIME_SERIES_ID.format(tsid);
```
Comment (martijnvg) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java:573]: 👍

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java [hunk 6]
```diff
@@ -108,7 +110,9 @@ public void writeTo(StreamOutput out) throws IOException {
         shardId.writeTo(out);
         out.writeStringArray(metrics);
         out.writeStringArray(labels);
-        out.writeOptionalStringArray(dimensions);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.DIMENSIONS_SERIALIZATION_PERSISTENT_TASK)) {
+            out.writeOptionalStringArray(dimensions);
+        }
```
Comment (martijnvg) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java:115]: Do we need to add a `if (dimensions.length != 0) {` check in the `toXContent(...)` method of this class? Just wondering if the xcontent format is shared between nodes?

**Review by martijnvg — APPROVED** [review]
I left one question otherwise LGTM

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle (hunk 4 above)
Comment (dnhatn) [reply x-pack/plugin/downsample/qa/mixed-cluster/build.gradle:8]: Yes, I will clean this up in a follow-up.

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java (hunk 6 above)
Comment (dnhatn) [reply x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardTaskParams.java:120]: ++, added a serializing test for this in https://github.com/elastic/elasticsearch/pull/106878/commits/7593cd8a410962cd64c7a0af0a2255708fc3dc56.

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  yamlRestTestImplementation project(path: xpackModule('rollup'))
+}
+
+restResources {
+  restApi {
+    include '_common', 'bulk', 'cluster', 'indices', 'search', 'ingest.put_pipeline', 'ingest.delete_pipeline'
+  }
+}
+
+artifacts {
```
Comment (mark-vieira) [inline x-pack/plugin/downsample/qa/mixed-cluster/build.gradle:30]: This block is only required if we plan to include these tests in the rest-resources artifact used by the clients team for their test. Given this is specifically for BWC testing I don't think it's applicable so we can remove this block.

File: x-pack/plugin/downsample/qa/mixed-cluster/src/javaRestTest/resources/tsdb-bulk-request.txt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+{"index": {}}
```
Comment (mark-vieira) [inline x-pack/plugin/downsample/qa/mixed-cluster/src/javaRestTest/resources/tsdb-bulk-request.txt:1]: What are these files needed for? I don't see any actual java rest test classes.

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  restXpackTests(new File(projectDir, "src/yamlRestTest/resources/rest-api-spec/test"))
+}
+
+
+def supportedVersion = bwcVersion -> {
+  return bwcVersion.onOrAfter(Version.fromString("8.8.0"));
+}
+
+BuildParams.bwcVersions.withWireCompatible(supportedVersion) { bwcVersion, baseName ->
+  def javaRestTest = tasks.register("v${bwcVersion}#javaRestTest", StandaloneRestIntegTestTask) {
```
Comment (mark-vieira) [inline x-pack/plugin/downsample/qa/mixed-cluster/build.gradle:40]: Perhaps I'm missing something but I don't see any java rest tests here.

Comment (dnhatn) [reply x-pack/plugin/downsample/qa/mixed-cluster/build.gradle]: I've removed this in https://github.com/elastic/elasticsearch/pull/106878/commits/5d095aec680af68c029d1bca4cee651c62fa5e10. I was adding a java rest test, but a yaml test should be enough here. Thanks for catching this.

File: x-pack/plugin/downsample/qa/mixed-cluster/src/javaRestTest/resources/tsdb-bulk-request.txt (hunk 8 above)
Comment (dnhatn) [reply x-pack/plugin/downsample/qa/mixed-cluster/src/javaRestTest/resources/tsdb-bulk-request.txt]: I've removed this in https://github.com/elastic/elasticsearch/pull/106878/commits/5d095aec680af68c029d1bca4cee651c62fa5e10

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle (hunk 7 above)
Comment (dnhatn) [reply x-pack/plugin/downsample/qa/mixed-cluster/build.gradle]: ++, I removed this in https://github.com/elastic/elasticsearch/pull/106878/commits/dfa59e9ba6666b4918a843cd2abd4f5e2ec08a74

File: x-pack/plugin/downsample/qa/mixed-cluster/build.gradle [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+restResources {
+  restApi {
+    include '_common', 'bulk', 'cluster', 'indices', 'search', 'ingest.put_pipeline', 'ingest.delete_pipeline'
+  }
+}
+
+def supportedVersion = bwcVersion -> {
+  return bwcVersion.onOrAfter(Version.fromString("8.8.0"));
```
Comment (mark-vieira) [inline x-pack/plugin/downsample/qa/mixed-cluster/build.gradle:29]: FWIW, this can be simplified to `bwcVersion.onOrAfter("8.8.0")`

**Review by mark-vieira — APPROVED** [review]

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Comment by dnhatn** [comment]
run elasticsearch-ci/part-1

**Comment by dnhatn** [comment]
Backported to 8.13 in https://github.com/elastic/elasticsearch/pull/106896

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java [hunk 11]
```diff
@@ -564,6 +564,15 @@ public XContentBuilder buildDownsampleDocument() throws IOException {
                 fieldProducer.write(builder);
             }
 
+            if (dimensions.length == 0) {
+                logger.debug("extracting dimensions from legacy tsid");
```
Comment (kkrik-es) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardIndexer.java:568]: I think there are two dimensions here:
 - Older index, using legacy tsid. This is controlled by IndexVersion.
 - Older transport layer, passing no dimensions. This is controlled by TransportVersion.

It seems possible to have a mix of older indexes with newer transport, as well as newer indexes with older transport. If so, can this fail in case we use a newer index (created in 8.13) with an updated tsid, through an older persistent task?
